src/librustc_mir/hair/cx/expr.rs

```diff
@@ -273,7 +273,8 @@ fn make_mirror_unadjusted<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
                             Def::VariantCtor(variant_id, CtorKind::Fn) => {
                                 Some((adt_def, adt_def.variant_index_with_id(variant_id)))
                             }
-                            Def::StructCtor(_, CtorKind::Fn) => Some((adt_def, 0)),
+                            Def::StructCtor(_, CtorKind::Fn) |
+                            Def::SelfCtor(..) => Some((adt_def, 0)),
```

Comment:
`SelfCtor` is not guaranteed to be `CtorKind::Fn`, sometimes it may be `CtorKind::Const`.
The constructor kind is determined during type checking when we find out whether `Self` actually refers to some constructor or not.

Reply:
There are few places in this commit in which `SelfCtor` is treated as `CtorKind::Fn`-only.

Reply:
Fixed.